reps/2022-04-21-object_store_HA.md

```diff
@@ -0,0 +1,132 @@
+
+In the chained object borrowing case, the owner of an object is not aware of the indirect borrowers. If all direct borrower fails, the owner will consider the object out-of-scope and GC the object. Accessing the object on indirect borrowers will fail with an `ObjectLostError`.
+
+more details: [issues 18456](https://github.com/ray-project/ray/issues/18456)
+
+#### Problem 3: Loss of All Copies
+
+Data of objects stored in the plasma store. For now, the plasma store is a thread of the raylet process, failure of the raylet process will lose data which store in plasma. Some objects which only one copy in that failed plasma store,  will be unavailable.
+
+### Proposed Design
+
+We implement
+#### Options to implement object HA with checkpoint
+
+We implement object HA based on the checkpoint, so we can walk around **Problem 3: Loss of All Copies**,
+previously discussed options: https://github.com/ray-project/enhancements/pull/10#issuecomment-1127719640
+
+We use highly available processes as global owners of checkpointed objects. Such highly available processes can be GCS or a group of named actors with `max_restarts=-1`. We reuse the existing ownership assignment RPCs to assign a checkpointed object to a global owner and encode the immutable info (an `owner_is_gcs` flag or the actor name) about the global owner into the owner address. The process to get an RPC client to the owner needs to be updated to be able to return a working RPC client to the up-to-date IP:port of the owner.
+
+Note that we don't need to restore the reference table in global owners by pulling info from the cluster because objects are already checkpointed. Checkpoint info is stored in the reference table and it will be encoded when serializing an object ref, hence checkpoint info is recoverable. If borrowers detected owner failure, they will try to reconnect to the owner and the recovered owner will recover the reference count and borrower list via these new RPC connections.
+
+- Pros
+  - No major protocol changes compared to the existing ownership assignment protocol.
+    - Low dev cost.
+  - No owner address updates because the `owner_is_gcs` flag or the actor name is encoded in it.
+    - Low dev cost.
+- Cons
+  - Centralized/semi-centralized architecture.
+    - Potentially bottleneck.
+    - Bad performance.
+  - Corner case handling such as RPC failures.
+    - Potentially high dev cost.
+
+We prefer named actors rather than GCS as global owners.
+
+- The number of global owners is configurable, hence scalable.
+- No need to embed (part of) core worker code into GCS.
+- No increased complexity for GCS.
+
+#### API:
+
+``` python
+# Set the number of global owner (default is zero) and the number of HA object's primary copies (default is zero).
+ray.init(
+    job_config=ray.job_config.JobConfig(
+        num_global_owners=16,
+        num_primary_copies=3,
+    )
+)
+
+# put a HA object. the default value of `enable_ha` is False.
+ray.put(value, enable_ha=True)
```

Comment:
I think there are use cases for specifying the exact actor that should own the object. In workflows, for example, this could be the WorkflowManagementActor.

Reply:
This part is not very clear on the REP, the previous consideration was to hope that this interface should be as simple as possible, user only needs to config `GlobalOwner` on `JobConfig` (like the number of `GlobalOwner`):
``` python
ray.init(job_config=JobConfig(global_owner={"number": 16}))
ray.put("test", enable_ha=True)
```

Maybe we can provide an interface to initialize `GlobalOwner`:
``` python
ray.init()
### creat owner [o1, o2, o3]
...
###
# or ray.init_global_owner(number=16)
ray.init_global_owner(owners=[o1, o2, o3])
ray.put("test", enable_ha=True)

```

Reply:
cc @kfstorm

Reply:
> I think there are use cases for specifying the exact actor that should own the object. In workflows, for example, this could be the WorkflowManagementActor.

Is there any reason that dummy actors as global owners don't work in workflows?